File: android/app/src/main/java/com/safering/android/domain/usecase/CheckCallUseCase.kt
package com.safering.android.domain.usecase

import com.safering.android.data.local.entity.CallLogEntity
import com.safering.android.data.repository.ScamRepository
import com.safering.android.domain.model.CallRisk
import javax.inject.Inject

/**
 * Use case for checking an incoming or outgoing call against the scam database.
 *
 * Steps:
 * 1. Hash the phone number (SHA-256)
 * 2. Check local Room DB cache
 * 3. If miss, check remote API
 * 4. Log the call to call_logs table
 * 5. Return CallRisk domain model
 *
 * Zero PII guarantee: Only the hash leaves the device.
 */
class CheckCallUseCase @Inject constructor(
    private val repository: ScamRepository
) {

    /**
     * Check a phone number and log the result.
     *
     * @param phoneNumber Raw phone number in E.164 format (e.g., "15551234567").
     * @param direction INCOMING or OUTGOING.
     * @param wasBlocked Whether the system blocked this call.
     * @return CallRisk assessment.
     */
    suspend operator fun invoke(
        phoneNumber: String,
        direction: String = CallLogEntity.DIRECTION_INCOMING,
        wasBlocked: Boolean = false
    ): CallRisk {
        val hash = repository.hashPhoneNumber(phoneNumber)

        val result = repository.checkNumber(phoneNumber)

        // Log the call
        repository.logCall(
            CallLogEntity(
                hash = hash,
                rawPrefix = phoneNumber.take(6),
                direction = direction,
                riskScore = result.risk,
                scamLabel = result.scamLabel,
                isScam = result.isHighRisk || result.isMediumRisk,
                wasBlocked = wasBlocked,
                timestamp = System.currentTimeMillis()
            )
        )

        return CallRisk(
            riskScore = result.risk,
            scamType = result.scamType,
            scamLabel = result.scamLabel,
            reportCount = result.reportCount,
            wasBlocked = wasBlocked,
            fromCache = result.fromCache,
            error = result.error
        )
    }
}

File: ios/SafeRing/Domain/UseCases/CheckCallUseCase.swift
import Foundation

/// Use case: Check an incoming or outgoing call for scam risk.
///
/// Steps:
/// 1. Hash the raw phone number using SHA-256.
/// 2. Query the local cache via repository (instant).
/// 3. If cache miss or stale, query the remote API.
/// 4. Optionally run the on-device ML number classifier.
/// 5. Return a `CallRisk` domain model.
///
/// # Zero PII
/// The raw phone number is hashed immediately and never leaves the device.
/// Only the SHA-256 hash is used for lookups.
///
final class CheckCallUseCase {

    // MARK: - Properties

    private let repository: ScamRepository
    private let numberClassifier: NumberClassifier?

    // MARK: - Initializer

    init(
        repository: ScamRepository,
        numberClassifier: NumberClassifier? = nil
    ) {
        self.repository = repository
        self.numberClassifier = numberClassifier
    }

    // MARK: - Execution

    /// Executes a full scam check on the given phone number.
    ///
    /// - Parameter rawNumber: The raw phone number string (e.g., "+15551234567").
    ///   This is immediately hashed — never persisted or transmitted in plaintext.
    /// - Returns: A CallRisk domain model with the assessment.
    /// - Throws: CheckCallError if the check cannot be completed.
    func execute(rawNumber: String) async throws -> CallRisk {
        // 1. Normalize and hash the phone number
        let normalized = normalizePhoneNumber(rawNumber)
        let hash = HashUtils.sha256(normalized)

        Logger.shared.info(
            "Checking call risk for hash: \(hash.prefix(8))...",
            category: .useCase
        )

        // 2. Build risk assessment
        let assessment: (score: Double, label: String?, confidence: Double, source: CallRisk.RiskSource)

        do {
            let result = try await repository.checkNumber(hash: hash)
            assessment = (
                score: result.riskScore,
                label: result.scamLabel,
                confidence: result.confidence,
                source: {
                    switch result.source {
                    case .localCache: return .localCache
                    case .remote:     return .remoteAPI
                    case .staleCache: return .staleCache
                    }
                }()
            )
        } catch {
            // 3. Fallback: try on-device ML classifier if available
            if let classifier = numberClassifier {
                let mlResult = classifier.classify(numberHash: hash)
                assessment = (
                    score: mlResult.riskScore,
                    label: mlResult.scamLabel,
                    confidence: mlResult.confidence,
                    source: .mlModel
                )
            } else {
                throw CheckCallError.checkFailed(underlying: error)
            }
        }

        // 4. Return domain model
        return CallRisk(
            hashedPhoneNumber: hash,
            score: assessment.score,
            confidence: assessment.confidence,
            scamType: assessment.label,
            isConfirmed: assessment.score >= AppConfig.autoBlockThreshold,
            shouldBlock: assessment.score >= AppConfig.autoBlockThreshold,
            source: assessment.source,
            assessedAt: Date()
        )
    }

    // MARK: - Helpers

    /// Normalizes a phone number to E.164 format for consistent hashing.
    /// Removes spaces, dashes, parentheses, and ensures + prefix.
    /// - Parameter number: Raw phone number string.
    /// - Returns: Normalized number string ready for hashing.
    private func normalizePhoneNumber(_ number: String) -> String {
        let digits = number.filter { $0.isNumber }
        // If number doesn't start with country code, assume +1 (US)
        if digits.hasPrefix("1") {
            return "+\(digits)"
        } else if digits.count == 10 {
            return "+1\(digits)"
        } else {
            return "+\(digits)"
        }
    }
}

// MARK: - Errors

enum CheckCallError: LocalizedError {
    case checkFailed(underlying: Error)

    var errorDescription: String? {
        switch self {
        case .checkFailed(let error):
            return "Call check failed: \(error.localizedDescription)"
        }
    }
}
